File: src/main/java/org/nest/nestml/_visitor/ConditionVisitor.java
package org.nest.nestml._visitor;

import de.se_rwth.commons.logging.Log;
import org.nest.nestml._ast.ASTExpr;
import org.nest.nestml._cocos.SplErrorStrings;
import org.nest.nestml._symboltable.typechecking.Either;
import org.nest.nestml._symboltable.symbols.TypeSymbol;
import org.nest.nestml._symboltable.unitrepresentation.UnitRepresentation;
import org.nest.utils.AstUtils;

import static org.nest.nestml._symboltable.typechecking.TypeChecker.isNumericPrimitive;
import static org.nest.nestml._symboltable.typechecking.TypeChecker.isUnit;
import static org.nest.nestml._symboltable.predefined.PredefinedTypes.getBooleanType;
import static org.nest.nestml._symboltable.predefined.PredefinedTypes.getRealType;

/**
 * @author ptraeder
 */
public class ConditionVisitor implements NESTMLVisitor {

  @Override
  public void visit(ASTExpr expr) {
    if (expr.getCondition().isPresent()) {
      final Either<TypeSymbol, String> condition = expr.getCondition().get().getType();
      final Either<TypeSymbol, String> ifTrueE = expr.getIfTrue().get().getType();
      final Either<TypeSymbol, String> ifNotE = expr.getIfNot().get().getType();

      if (condition.isError()) {
        expr.setType(condition);
        return;
      }
      if (ifTrueE.isError()) {
        expr.setType(ifTrueE);
        return;
      }
      if (ifNotE.isError()) {
        expr.setType(ifNotE);
        return;
      }

      TypeSymbol ifTrue = ifTrueE.getValue();
      TypeSymbol ifNot = ifNotE.getValue();

      if (!condition.getValue().equals(getBooleanType())) {
        final String errorMsg = CommonsErrorStrings.messageTernary(this, expr.get_SourcePositionStart());
        expr.setType(Either.error(errorMsg));
        Log.error(errorMsg, expr.get_SourcePositionStart());
        return;
      }
      //Alternatives match exactly -> any is valid
      if (ifTrue.prettyPrint().equals(ifNot.prettyPrint())) {
        expr.setType(Either.value(ifTrue));
        return;
      }

      //Both are units -> try to recover, otherwise real
      if(isUnit(ifTrue)&&isUnit(ifNot)){
        UnitRepresentation ifNotRep = UnitRepresentation.getBuilder().serialization(ifTrue.getName()).build();
        UnitRepresentation ifTrueRep = UnitRepresentation.getBuilder().serialization(ifNot.getName()).build();
        if(ifTrueRep.equalBase(ifNotRep)) { //matching base, recover
          //Determine the difference in magnitude
          int magDiff = ifNotRep.getMagnitude() - ifTrueRep.getMagnitude();

          //replace left expression with multiplication
          expr.setIfTrue(AstUtils.createSubstitution(expr.getIfTrue().get(),magDiff));

          //revisit current sub-tree with substitution
          ExpressionTypeVisitor expressionTypeVisitor = new ExpressionTypeVisitor();
          expr.accept(expressionTypeVisitor);

          //drop warning about implicit conversion
          final String warnMsg = SplErrorStrings.messageImplicitConversion(this,
              ifTrueRep.prettyPrint(),
              ifNotRep.prettyPrint(),
              expr.get_SourcePositionStart());
          Log.warn(warnMsg,expr.get_SourcePositionStart());
          return;
        }


        final String errorMsg = CommonsErrorStrings.messageTrueNot(
            this,
            ifTrue.prettyPrint(),
            ifNot.prettyPrint(),
            "real",
            expr.get_SourcePositionStart());
        expr.setType(Either.value(getRealType()));
        Log.warn(errorMsg, expr.get_SourcePositionStart());
        return;
      }
      //one Unit and one numeric primitive and vice versa -> assume unit,warn
      if ((isUnit(ifTrue) && isNumericPrimitive(ifNot)) ||
          isUnit(ifNot) && isNumericPrimitive(ifTrue)) {
        TypeSymbol unitType;
        if (isUnit(ifTrue)) {
          unitType = ifTrue;
        }
        else {
          unitType = ifNot;
        }
        final String errorMsg = CommonsErrorStrings.messageTrueNot(
            this,
            ifTrue.prettyPrint(),
            ifNot.prettyPrint(),
            unitType.prettyPrint(),
            expr.get_SourcePositionStart());
        expr.setType(Either.value(unitType));
        Log.warn(errorMsg, expr.get_SourcePositionStart());
        return;
      }

      //both are numeric primitives (and not equal) ergo one is real and one is integer -> real
      if (isNumericPrimitive(ifTrue) && isNumericPrimitive(ifNot)) {
        expr.setType(Either.value(getRealType()));
        return;
      }

      //if we get here it is an error
      final String errorMsg = CommonsErrorStrings.messageTypeMissmatch(
          this,
          ifTrue.prettyPrint(),
          ifNot.prettyPrint(),
          expr.get_SourcePositionStart());
      expr.setType(Either.error(errorMsg));
      Log.error(errorMsg, expr.get_SourcePositionStart());
    }

  }

}

File: pynestml/modelprocessor/ConditionVisitor.py
#
# ConditionVisitor.py
#
# This file is part of NEST.
#
# Copyright (C) 2004 The NEST Initiative
#
# NEST is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# NEST is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with NEST.  If not, see <http://www.gnu.org/licenses/>.

"""
expression : condition=expression '?' ifTrue=expression ':' ifNot=expression
"""
from copy import copy

from pynestml.modelprocessor.ASTExpression import ASTExpression
from pynestml.modelprocessor.Either import Either
from pynestml.modelprocessor.ErrorStrings import ErrorStrings
from pynestml.modelprocessor.ErrorTypeSymbol import ErrorTypeSymbol
from pynestml.modelprocessor.ModelVisitor import NESTMLVisitor
from pynestml.modelprocessor.PredefinedTypes import PredefinedTypes
from pynestml.modelprocessor.UnitTypeSymbol import UnitTypeSymbol
from pynestml.utils.Logger import Logger, LOGGING_LEVEL
from pynestml.utils.Messages import MessageCode


class ConditionVisitor(NESTMLVisitor):
    """
    This visitor is used to derive the correct type of a ternary operator, i.e., of all its subexpressions.
    """

    def visit_expression(self, _expr=None):
        """
        Visits an expression consisting of the ternary operator and updates its type.
        :param _expr: a single expression
        :type _expr: ASTExpression
        """
        if_true = _expr.getIfTrue().type
        if_not = _expr.getIfNot().type
        condition = _expr.getCondition().type

        condition.referenced_object = _expr.getCondition()
        if_true.referenced_object = _expr.getIfTrue()
        if_not.referenced_object = _expr.getIfNot()

        # Condition must be a bool
        if not condition.equals(PredefinedTypes.getBooleanType()):
            error_msg = ErrorStrings.messageTernary(self, _expr.getSourcePosition())
            _expr.type = ErrorTypeSymbol()
            Logger.logMessage(_message=error_msg, _errorPosition=_expr.getSourcePosition(),
                              _code=MessageCode.TYPE_DIFFERENT_FROM_EXPECTED,
                              _logLevel=LOGGING_LEVEL.ERROR)
            return

        # Alternatives match exactly -> any is valid
        if if_true.equals(if_not):
            _expr.type = copy(if_true)
            return

        # Both are units but not matching-> real WARN
        if isinstance(if_true, UnitTypeSymbol) and isinstance(if_not, UnitTypeSymbol):
            # TODO: This is not covered by our tests, and it doesnt work
            error_msg = ErrorStrings.messageTernaryMismatch(self, if_true.print_symbol(), if_not.print_symbol(),
                                                            _expr.getSourcePosition())
            _expr.type = PredefinedTypes.getRealType()
            Logger.logMessage(_message=error_msg,
                              _code=MessageCode.TYPE_DIFFERENT_FROM_EXPECTED,
                              _errorPosition=_expr.getIfTrue().getSourcePosition(),
                              _logLevel=LOGGING_LEVEL.WARNING)
            return

        # one Unit and one numeric primitive and vice versa -> assume unit, WARN
        if (isinstance(if_true, UnitTypeSymbol) and if_not.isNumericPrimitive()) or (
                    isinstance(if_not, UnitTypeSymbol) and if_true.isNumericPrimitive()):

            if isinstance(if_true, UnitTypeSymbol):
                unit_type = if_true
            else:
                unit_type = if_not
            error_msg = ErrorStrings.messageTernaryMismatch(self, str(if_true), str(if_not),
                                                            _expr.getSourcePosition())
            _expr.type = copy(unit_type)
            Logger.logMessage(_message=error_msg,
                              _code=MessageCode.TYPE_DIFFERENT_FROM_EXPECTED,
                              _errorPosition=_expr.getIfTrue().getSourcePosition(),
                              _logLevel=LOGGING_LEVEL.WARNING)
            return

        # both are numeric primitives (and not equal) ergo one is real and one is integer -> real
        if if_true.isNumericPrimitive() and if_not.isNumericPrimitive():
            _expr.type = PredefinedTypes.getRealType()
            return

        # if we get here it is an error
        error_msg = ErrorStrings.messageTernaryMismatch(self, str(if_true), str(if_not),
                                                        _expr.getSourcePosition())
        _expr.type = ErrorTypeSymbol()
        Logger.logMessage(_message=error_msg,
                          _errorPosition=_expr.getSourcePosition(),
                          _code=MessageCode.TYPE_DIFFERENT_FROM_EXPECTED,
                          _logLevel=LOGGING_LEVEL.ERROR)
